### DISCUSSION TITLE
Is it a good idea to add HX-Trigger-After-Swap/Settle:DomContentLoaded response header if whole body swapped?

### DISCUSSION THREAD

**Discussion by mic4126** [bigskysoftware/htmx#1786] [General]
Currently I working on an internal old jsp project with a bunch of jquery script.  

After swapping out whold body tag, I found only possible way to add all event listener back (without rewriting a bunch of jquery js) is to fire DomContentLoaded by htmx. 

and I think a little bit weird to fire native event by myself.

**Comment by Telroshan** [comment]
Indeed, manually firing `DomContentLoaded` sounds weird. To be honest, I didn't even know this would work 😆 

Assuming you're talking about your own event listeners here, and not the ones that htmx sets up internally, you could use [mutation observers](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) to listen for added nodes to the DOM, and bind listeners to them

**Comment by mic4126** [comment]
In my working project, almost event listener all are added by jQuery.on using jQuery.ready (except new js written by me)

like $($("#btn").on(....))

https://api.jquery.com/on/
http://learn.jquery.com/using-jquery-core/document-ready/ 

so I need someway to trigger all event listener add by jQuery and obviously I find an weird but at least working way.

Maybe better to let just trigger all jQuery ready event listener?
(maybe by hacky way)

**Reply by Telroshan** [reply]
I mean, as long as it works, that's fine!

I often have the situation on my projects, where I have some special "components" in some pages, that are sometimes on the initial page, and some other times get swapped in specific places during the page's lifetime.

The approach I ended up using, is a function both called on load and on mutations, such as
```javascript
/** @param {HTMLElement} root
 *  @param {string} selector
 *  @param {(HTMLElement)=>void} callback */
function setupRootOrChildren(root, selector, callback) {
    if (root.matches(selector)) {
        callback(root)
    }
    root.querySelectorAll(selector).forEach(callback)
}

/** @param {HTMLElement} root */
function setupElements(root) {
    setupRootOrChildren(root, "[some-custom-attribute]", function (element) {
        // bind listeners on element
    })
    setupRootOrChildren(root, ".some-class", function (element) {
        // bind listeners on element
    })
    setupRootOrChildren(root, "#aSpecificID", function (element) {
        // bind listeners on element
    })
}

document.addEventListener("DOMContentLoaded", function () {
    setupElements(document.body)

    new MutationObserver(function (mutations) {
        mutations.forEach(function (mutation) {
            mutation.addedNodes.forEach(function (node) {
                if (node.nodeType === Node.ELEMENT_NODE) {
                    setupElements(node)
                }
            })
        })
    }).observe(document.body, {
        subtree: true,
        childList: true,
    })
})
```

- See here how the listeners are bound on page load thanks to `DOMContentLoaded` + `setupElements(document.body)`
- Also how any element swapped in the page would get setup through the `MutationObserver`, for example swapping a `<input some-custom-attribute>` would be caught.

That approach is more complex than simply firing the `DOMContentLoaded`, it adds more code.
The advantage though is that it would support any mutation applied to the page (so swapping specific elements and not only the whole body), automatically.

But again, if you don't need that and are only swapping the whole body anyway, firing `DOMContentLoaded` sounds perfectly fine!

**Reply by mic4126** [reply]
This solution is great. However, I do not have the bravery to rewrite all existing code. 😢

**Comment by mic4126** [comment]
Just found that jQuery store all event listener in itself.
https://stackoverflow.com/a/447106

seems possible to write a plugin to fire all jQuery event listener
